doc/adr/2022-11-16-option-precedence-hierarchy.md

```diff
@@ -0,0 +1,28 @@
+**Title**: Establish how multiple values for the same option are reconciled.
+
+**Status**: proposed
+
+**Context**: There are a variety of ways that options can be customized for a
+given rpc call. Both `Client` and `Connection` classes can be created with a set
+of `google::cloud::Options`. Individual function calls in a `Client` can be
+passed `google::cloud::Options`. Some environment variables can be set that
+such that the variable value sets an Option associated with that environment
+variable. Lastly, GCS `Request` types inherit from various Option templates
+which allow the corresponding option to be set in an instance of the `Request`.
+
+When the same Option is set, with different values, via more than one of these
+mechanisms, there needs to be a uniform process of determining which value to
+use.
+
+**Decision**: The following hierarchy will be followed when determining which
+value to use when the same Option is set via multiple mechanisms:
+
+1. Environment variables
+1. Per function call/GCS `Request` instance (if both are set with different values, an error is returned)
```

Comment:
What about just:
```
1. Operation Options
```

then a section later that says:
```
In GCS, there are multiple ways to set the same option. For example,
a customer can supply `google::cloud::UserIpOption` in a
`google::cloud::Options` or supply `google::cloud::storage::IpOption`
directly to a `storage::Client::Foo(..., Options&&)` parameter pack. If
the same option is set to different values, an error is returned.
```

Also, in GCS, >1 `gc::Options` can be supplied to the parameter pack. I think we merge them together, with the first one in the list being preferred. I doubt this is worth calling out in the ADR though.

Reply:
> I'd rather have the more terse verbiage in line than split into two sections.

I claim that saying: "Function `Options`" is the most terse verbiage, and captures the intent best.

GCS `Request` options are just one type of function options. (Yes they are weird and deserve a call out. But I don't think they belong in-line).

If you are unconvinced, so be it.